cs/src/Connections/TunnelClient.cs

```diff
@@ -254,6 +258,43 @@ private void ConfigurePortForwardingService()
         {
             pfs.MessageFactory = this;
             pfs.ForwardedPortConnecting += OnForwardedPortConnecting;
+            pfs.RemoteForwardedPorts.PortAdded += (_, e) => OnForwardedPortAdded(pfs, e);
+            pfs.RemoteForwardedPorts.PortUpdated += (_, e) => OnForwardedPortAdded(pfs, e);
+        }
+    }
+
+    private void OnForwardedPortAdded(PortForwardingService pfs, ForwardedPortEventArgs e)
+    {
+        var port = e.Port.RemotePort;
+        if (!port.HasValue)
+        {
+            return;
+        }
+
+        var disconnectedStreamsCount = 0;
+        lock (this.disconnectedStreams)
+        {
+            // If there are disconnected streams for the port, re-connect them now.
+            if (this.disconnectedStreams.TryGetValue(port.Value, out var streams))
+            {
+                disconnectedStreamsCount = streams.Count;
+            }
+        }
+
+        for (int i = 0; i < disconnectedStreamsCount; i++)
+        {
+            Task.Run(async () =>
+            {
+                try
+                {
+                    await pfs.ConnectToForwardedPortAsync(port.Value, CancellationToken.None);
```

Comment:
I'm not too familiar with how channels work - do they not have a name or ID? I'm just wondering why it's not necessary to pass the value of `i` or `streams[i]`?

Reply:
SSH channels have an integer ID, however we don't store a mapping from channel ID to `SecureStream` instance because the channel IDs may be different after the reconnection.

The SSH session inside `SecureStream` has a (cryptographically strong) ID that is used to match the streams when reconnecting, as commented on line 332.